File: durga-runtime/src/main/java/org/gautelis/durga/plugins/KvEnricher.java
package org.gautelis.durga.plugins;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;

import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

/**
 * Enriches a JSON payload by looking up external data and merging it in.
 *
 * <p>Supports two source modes:
 * <ul>
 * <li>{@code inline} — data is provided directly in configuration as a map of key to
 *     JSON objects</li>
 * <li>{@code kafka_topic} — (future) data is looked up from a compacted Kafka topic</li>
 * </ul>
 *
 * <p>The enrichment is keyed by the value of {@code keyField} in the input payload.
 */
public final class KvEnricher implements Plugin {

    @Override
    public String execute(String payload, String config) throws Exception {
        String keyField = "_id";
        java.util.Map<String, String> inline = new java.util.LinkedHashMap<>();
        if (config != null && !config.isBlank()) {
            String[] parts = config.split(";");
            for (String part : parts) {
                int eq = part.indexOf('=');
                if (eq <= 0) {
                    continue;
                }
                String key = part.substring(0, eq).trim();
                String val = part.substring(eq + 1).trim();
                switch (key) {
                    case "keyField" -> keyField = val;
                    case "inline" -> {
                        if (val.startsWith("{") && val.endsWith("}")) {
                            inline.putAll(parseInlineMap(val.substring(1, val.length() - 1)));
                        }
                    }
                }
            }
        }
        KvEnricher enricher = new KvEnricher(keyField, inline);
        return enricher.enrich(payload);
    }

    /**
     * Parses the inline map body ({@code key:{json}, key:{json}}) into a map of key to raw
     * enrichment JSON. Splitting is brace/bracket/quote aware, so commas and colons inside a
     * value's JSON object do not break the entry boundaries.
     */
    static Map<String, String> parseInlineMap(String body) {
        Map<String, String> map = new java.util.LinkedHashMap<>();
        int depth = 0;
        boolean inString = false;
        char quote = 0;
        int entryStart = 0;
        int colon = -1;
        for (int i = 0; i < body.length(); i++) {
            char c = body.charAt(i);
            if (inString) {
                if (c == quote && (i == 0 || body.charAt(i - 1) != '\\')) {
                    inString = false;
                }
                continue;
            }
            switch (c) {
                case '"', '\'' -> { inString = true; quote = c; }
                case '{', '[' -> depth++;
                case '}', ']' -> depth--;
                case ':' -> { if (depth == 0 && colon < 0) colon = i; }
                case ',' -> {
                    if (depth == 0) {
                        addInlineEntry(map, body, entryStart, colon, i);
                        entryStart = i + 1;
                        colon = -1;
                    }
                }
                default -> { }
            }
        }
        addInlineEntry(map, body, entryStart, colon, body.length());
        return map;
    }

    private static void addInlineEntry(Map<String, String> map, String body, int start, int colon, int end) {
        if (colon < 0 || colon <= start) {
            return;
        }
        String key = body.substring(start, colon).trim();
        String value = body.substring(colon + 1, end).trim();
        if (!key.isEmpty() && !value.isEmpty()) {
            map.put(key, value);
        }
    }

    private final String keyField;
    private final Map<String, String> inlineData;

    public KvEnricher() {
        this("_id", Map.of());
    }

    public KvEnricher(String keyField, Map<String, String> inlineData) {
        this.keyField = keyField;
        this.inlineData = inlineData != null
                ? new ConcurrentHashMap<>(inlineData)
                : new ConcurrentHashMap<>();
    }

    /**
     * Enriches a JSON payload by looking up the key field value in the inline data store
     * and shallow-merging any matching enrichment data.
     *
     * @param json input JSON string
     * @return enriched JSON string
     */
    public String enrich(String json) {
        ObjectMapper mapper = new ObjectMapper();
        JsonNode input;
        try {
            input = mapper.readTree(json);
        } catch (JsonProcessingException e) {
            throw new IllegalArgumentException("Invalid input JSON", e);
        }

        JsonNode keyNode = PipelinePlugin.fieldAt(input, keyField);
        if (keyNode == null || keyNode.isNull()) {
            return json;
        }
        String key = keyNode.asText();
        String enrichmentJson = inlineData.get(key);
        if (enrichmentJson == null) {
            return json;
        }

        JsonNode enrichment;
        try {
            enrichment = mapper.readTree(enrichmentJson);
        } catch (JsonProcessingException e) {
            throw new IllegalArgumentException("Invalid enrichment JSON for matching key", e);
        }

        return PipelinePlugin.shallowMerge(input, enrichment).toString();
    }

    /**
     * Returns the key field name used for lookups.
     */
    public String keyField() {
        return keyField;
    }

    /**
     * Returns an unmodifiable view of the inline enrichment data.
     */
    public Map<String, String> inlineData() {
        return Map.copyOf(inlineData);
    }
}

File: durga-rust/src/plugins/kv_enricher.rs
//! Port of `org.gautelis.durga.plugins.KvEnricher` (plugin id `kv-enricher`).
//!
//! Looks up a payload key field in an inline map and shallow-merges the
//! matching enrichment JSON. The inline map is parsed brace/quote aware so
//! enrichment objects containing commas are not shredded.

use std::collections::HashMap;

use serde_json::Value;

use crate::pipeline::{field_at, shallow_merge};
use crate::plugin::{Plugin, PluginError};

pub struct KvEnricher {
    key_field: String,
    inline: HashMap<String, String>,
}

impl KvEnricher {
    pub fn new() -> Self {
        KvEnricher { key_field: "_id".to_string(), inline: HashMap::new() }
    }

    fn from_config(config: &str) -> Self {
        let mut key_field = "_id".to_string();
        let mut inline = HashMap::new();
        for part in config.split(';') {
            let Some(eq) = part.find('=') else { continue };
            if eq == 0 {
                continue;
            }
            let key = part[..eq].trim();
            let val = part[eq + 1..].trim();
            match key {
                "keyField" => key_field = val.to_string(),
                "inline" if val.starts_with('{') && val.ends_with('}') => {
                    inline.extend(parse_inline_map(&val[1..val.len() - 1]));
                }
                _ => {}
            }
        }
        KvEnricher { key_field, inline }
    }

    /// Enriches `json` by looking up the key field in the inline map and
    /// shallow-merging the matching enrichment object. Passes through unchanged
    /// when the key is missing or unmatched.
    pub fn enrich(&self, json: &str) -> Result<String, PluginError> {
        let input: Value = serde_json::from_str(json)
            .map_err(|e| -> PluginError { format!("Invalid input JSON: {e}").into() })?;
        let key = match field_at(&input, &self.key_field) {
            Some(v) if !v.is_null() => match v {
                Value::String(s) => s.clone(),
                other => other.to_string(),
            },
            _ => return Ok(json.to_string()),
        };
        let Some(enrichment_json) = self.inline.get(&key) else {
            return Ok(json.to_string());
        };
        let enrichment: Value = serde_json::from_str(enrichment_json).map_err(|e| -> PluginError {
            format!("Invalid enrichment JSON for matching key: {e}").into()
        })?;
        Ok(shallow_merge(&input, &enrichment).to_string())
    }
}

impl Default for KvEnricher {
    fn default() -> Self {
        Self::new()
    }
}

/// Parses the inline map body (`key:{json}, key:{json}`) into key -> raw JSON.
/// Splitting is brace/bracket/quote aware so commas inside a value's JSON do
/// not break entry boundaries.
pub fn parse_inline_map(body: &str) -> HashMap<String, String> {
    let mut map = HashMap::new();
    let bytes = body.as_bytes();
    let mut depth: i32 = 0;
    let mut in_string = false;
    let mut quote = 0u8;
    let mut entry_start = 0usize;
    let mut colon: isize = -1;
    let mut i = 0usize;
    while i < bytes.len() {
        let c = bytes[i];
        if in_string {
            if c == quote && (i == 0 || bytes[i - 1] != b'\\') {
                in_string = false;
            }
            i += 1;
            continue;
        }
        match c {
            b'"' | b'\'' => {
                in_string = true;
                quote = c;
            }
            b'{' | b'[' => depth += 1,
            b'}' | b']' => depth -= 1,
            b':' if depth == 0 && colon < 0 => colon = i as isize,
            b',' if depth == 0 => {
                add_entry(&mut map, body, entry_start, colon, i);
                entry_start = i + 1;
                colon = -1;
            }
            _ => {}
        }
        i += 1;
    }
    add_entry(&mut map, body, entry_start, colon, body.len());
    map
}

fn add_entry(map: &mut HashMap<String, String>, body: &str, start: usize, colon: isize, end: usize) {
    if colon < 0 || (colon as usize) <= start {
        return;
    }
    let colon = colon as usize;
    let key = body[start..colon].trim();
    let value = body[colon + 1..end].trim();
    if !key.is_empty() && !value.is_empty() {
        map.insert(key.to_string(), value.to_string());
    }
}

impl Plugin for KvEnricher {
    fn execute_str(&self, payload: &str, config: &str) -> Result<Option<String>, PluginError> {
        Ok(Some(Self::from_config(config).enrich(payload)?))
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn enriches_matching_key_with_multifield_object() {
        let p = KvEnricher::new();
        let out = p
            .execute_str(
                r#"{"customer_email":"bob@example.com","amount":900}"#,
                "keyField=customer_email;inline={alice@example.com:{\"tier\":\"gold\",\"rep\":\"Alice\"}, bob@example.com:{\"tier\":\"silver\",\"rep\":\"Bob\"}}",
            )
            .unwrap()
            .unwrap();
        let v: Value = serde_json::from_str(&out).unwrap();
        assert_eq!(v["tier"], "silver");
        assert_eq!(v["rep"], "Bob");
        assert_eq!(v["amount"], 900);
    }

    #[test]
    fn passes_through_on_missing_key() {
        let p = KvEnricher::new();
        let input = r#"{"customer_email":"nobody@example.com"}"#;
        let out = p
            .execute_str(input, "keyField=customer_email;inline={a:{\"x\":1}}")
            .unwrap()
            .unwrap();
        assert_eq!(out, input);
    }

    #[test]
    fn parse_inline_respects_brace_depth() {
        let m = parse_inline_map("a:{\"x\":1,\"y\":2}, b:{\"z\":3}");
        assert_eq!(m.len(), 2);
        assert_eq!(m.get("a").unwrap(), "{\"x\":1,\"y\":2}");
        assert_eq!(m.get("b").unwrap(), "{\"z\":3}");
    }
}
